### PR TITLE
Part 3

### PR DESCRIPTION
**Pull request by davidtabat**

### CODE REVIEW

File: Devall/Tabatadze/Api/CompanyRepositoryInterface.php [hunk 1]
```diff
@@ -2,21 +2,29 @@
...
+    public function getByIdApi($id);
+
+    /**
+     * @return Collection
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
      */
-    public function getList(SearchCriteriaInterface $searchCriteria);
+    public function getList();
+
+
```
Comment (bricht) [inline Devall/Tabatadze/Api/CompanyRepositoryInterface.php:29]: Why dont you also implement save and delete? I know you dont use them for now, but the repository is responsible for CRUD operations, so it would be weird to create a repository and only implement half of them.

File: Devall/Tabatadze/Api/Data/CompanyInterface.php [hunk 2]
```diff
@@ -13,4 +13,24 @@ public function getId();
      * @return string
      */
     public function getName();
+
```
Comment (bricht) [inline Devall/Tabatadze/Api/Data/CompanyInterface.php:16]: Here also should implement set methods for all the attributes, like you would expect from any model.

File: Devall/Tabatadze/Api/CompanyRepositoryInterface.php [hunk 3]
```diff
@@ -2,21 +2,29 @@
...
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function getByIdApi($id);
+
+    /**
+     * @return Collection
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
      */
-    public function getList(SearchCriteriaInterface $searchCriteria);
+    public function getList();
```
Comment (bricht) [inline Devall/Tabatadze/Api/CompanyRepositoryInterface.php:27]: Why did you delete the SearchCriteriaInterface?

File: Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+use \Magento\Framework\View\Element\Template;
+use Devall\Tabatadze\Model\CompanyRepository;
+use Magento\Customer\Model\Session as CustomerSession;
+use Magento\Customer\Model\Customer;
+
+class CompanyBlock extends Template
+{
+    public function __construct(
+        CustomerSession $customerSession,
+        CompanyRepository $companyRepository,
```
Comment (bricht) [inline Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php:13]: Use the CompanyRepositoryInterface instead of the repository directly. 
Its the D of SOLID principles aka Service Contracts in Magento (https://devdocs.magento.com/guides/v2.4/extension-dev-guide/service-contracts/service-contracts.html)

File: Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        parent::__construct($context, $data);
+        $this->customer = $customer;
+        $this->customerSession = $customerSession;
+        $this->companyRepository = $companyRepository;
+    }
+
+    public function hasCompany(){
+        $customerId = $this->getCustomerId();
+        $customer = $this->customer->load($customerId);
```
Comment (bricht) [inline Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php:26]: Dont use Customer model to load itself, instead use CustomerRepositoryInterface.

File: Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Template\Context $context,
+        array $data = [])
+    {
+        parent::__construct($context, $data);
+        $this->customer = $customer;
+        $this->customerSession = $customerSession;
+        $this->companyRepository = $companyRepository;
+    }
+
+    public function hasCompany(){
```
Comment (bricht) [inline Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php:24]: Add docBlocks everywhere. That way we know what parameters we expect to pass to your functions and what return values we expect.

File: Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+{
+    public function __construct(
+        CustomerSession $customerSession,
+        CompanyRepository $companyRepository,
+        Customer $customer,
+        Template\Context $context,
+        array $data = [])
+    {
+        parent::__construct($context, $data);
+        $this->customer = $customer;
```
Comment (bricht) [inline Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php:19]: You are declaring the properties dynamically. Not good. Declare the properties at the start of the class.

File: Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        $this->customerSession = $customerSession;
+        $this->companyRepository = $companyRepository;
+    }
+
+    public function hasCompany(){
+        $customerId = $this->getCustomerId();
+        $customer = $this->customer->load($customerId);
+        $customerData = $customer->getDataModel();
+        $companyId = $customerData->getCustomAttribute('customer_pan_number');
```
Comment (bricht) [inline Devall/Tabatadze/Block/Account/Dashboard/CompanyBlock.php:28]: If you are hardcoding a value like this it is good to put in a constant at the start of the class. Or even better define it in one class somewhere, like the model and then reference that constant everywhere you need this value.
something like:
```
/**
 * Customer attribute code
 */
private const COMPANY_ATTRIBUTE_CODE = 'customer_pan_number';

...

$customerData->getCustomAttribute(self::COMPANY_ATTRIBUTE_CODE);
```

File: Devall/Tabatadze/Block/CompanyList.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+{
+    /**
+     * @var Collection
+     */
+    private $collection;
+
+    /**
+     * Display constructor.
+     * @param Template\Context $context
+     * @param Collection $collection
```
Comment (bricht) [inline Devall/Tabatadze/Block/CompanyList.php:20]: Incomplete docBlock

File: Devall/Tabatadze/Block/CompanyList.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        CompanyRepository $companyRepository,
+        Customer $customer,
+        Template\Context $context,
+        Collection $collection,
+        array $data = []
+    ) {
+        $this->collection = $collection;
+        parent::__construct($context, $data);
+        $this->customer = $customer;
+        $this->customerSession = $customerSession;
```
Comment (bricht) [inline Devall/Tabatadze/Block/CompanyList.php:34]: Dynamically declared properties. You need to configure your phpStorm to notify you of these things. Helps to notice these problems and also correct automatically.
image

File: Devall/Tabatadze/Block/CompanyList.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        $this->companyRepository = $companyRepository;
+    }
+
+    public function getCompanies(){
```
Comment (bricht) [inline Devall/Tabatadze/Block/CompanyList.php:38]: No docBlock here and in other functions in this class

File: Devall/Tabatadze/Block/CompanyList.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        return $this->customerSession->getId();
+    }
+
+    public function compareCompany(){
+        $customerId = $this->getCustomerId();
+        $customer = $this->customer->load($customerId);
+        $customerData = $customer->getDataModel();
+        $companyId = $customerData->getCustomAttribute('customer_pan_number');
+        if (!empty($companyId)) {
```
Comment (bricht) [inline Devall/Tabatadze/Block/CompanyList.php:62]: Configure phpStorm and it will tell you what the problem here is :P, its not a big deal, but still. Code quality is super important.

File: Devall/Tabatadze/Block/CompanyList.php [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+            return $companyId->getValue();
+        }
+        return false;
+    }
+
+    public function getCompany(){
+        $customerId = $this->getCustomerId();
+        $customer = $this->customer->load($customerId);
```
Comment (bricht) [inline Devall/Tabatadze/Block/CompanyList.php:70]: Use repositories, dont use models to load themselves. The method is deprecated and a leftover from magento 1. 
image
image

File: Devall/Tabatadze/Console/Command/CompanyTest.php [hunk 14]
```diff
@@ -33,20 +40,19 @@ protected function configure()
...
      * @param InputInterface $input
      * @param OutputInterface $output
      *
      * @return void
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
         $output->writeln('<info>Testing Companies.</info>');
-        $repo = $this->objectManager->get('Devall\Tabatadze\Model\CompanyRepository');
```
Comment (bricht) [inline Devall/Tabatadze/Console/Command/CompanyTest.php:47]: I hope you know not to use object manager in your final code, but in this case it is a test class for yourself, that you would not ship to a client, so I guess it doesnt matter.

File: Devall/Tabatadze/Controller/Company/EditPost.php [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        $this->customer = $customer;
+        $this->session = $customerSession;
+        $this->resultPageFactory = $resultPageFactory;
+        parent::__construct($context);
+    }
+
+    public function execute()
+    {
+        try {
+            $data = $this->getRequest()->getParam('company');
```
Comment (bricht) [inline Devall/Tabatadze/Controller/Company/EditPost.php:36]: Best to put in a constant if its a hardcoded value like this.

File: Devall/Tabatadze/Controller/Company/EditPost.php [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+                $customerData->setCustomAttribute('customer_pan_number',$data);
+                $customer->updateData($customerData);
+                $customer->save();
+                $this->messageManager->addSuccessMessage(__("Data Saved Successfully."));
+            }
+        } catch (Exception $e) {
+            $this->messageManager->addErrorMessage($e, __("We can\'t submit your request, Please try again."));
+        }
+        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+        $resultRedirect->setUrl('http://localhost/customer/account');
```
Comment (bricht) [inline Devall/Tabatadze/Controller/Company/EditPost.php:50]: hardcoded localhost? nooo

File: Devall/Tabatadze/Controller/Company/EditPost.php [hunk 17]
```diff
@@ -0,0 +1,54 @@
+<?php
+
```
Comment (bricht) [inline Devall/Tabatadze/Controller/Company/EditPost.php:2]: Check out \Magento\Customer\Controller\Account\EditPost
I think there are a few things there for security/improved performance that you can implement like validating formKey and disabling address validation.

File: Devall/Tabatadze/Model/CompanyRepository.php [hunk 18]
```diff
@@ -20,79 +14,35 @@ class CompanyRepository implements CompanyRepositoryInterface
     public $companyFactory;
 
     /**
-     * @var CompanyCollectionFactory
+     * @var Company
```
Comment (bricht) [inline Devall/Tabatadze/Model/CompanyRepository.php:17]: Yeea, there is so little here theres nothing to review :D. You dont want to leave unused code in blocks, viewmodels etc, but you should implement the repository fully.
Also please do create another pull request to an empty branch. This is really hard to read.

File: Devall/Tabatadze/view/frontend/templates/companyinfo.phtml [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+/** @var \Devall\Tabatadze\Block\CompanyList $block */
+
+$companies = $block->getCompanies();
+?>
+<form method="post" id="companyform" action="<?= $block->escapeUrl($block->getAction()) ?>">
+    <?= $block->getBlockHtml('formkey') ?>
+    <select name="company" id="company">
+        <?php foreach ($companies as $company): ?>
+            <option
+                value="<?= $company->getId() ?>" <?php if ($company->getId() == $block->compareCompany()): ?> selected <?php endif; ?>>
```
Comment (bricht) [inline Devall/Tabatadze/view/frontend/templates/companyinfo.phtml:11]: I would exchange 
$block->compareCompany()
to 
$block->getCustomerCompanyId() or something like that. Its not clear what compareCompany() is .

File: Devall/Tabatadze/view/frontend/web/js/details.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+], function (Component, ko, $) {
+    return Component.extend({
+        initialize: function (config) {
+            var feature = this;
+            this._super();
+            this.country(config.countrycompany);
+            this.street(config.streetcompany);
+            this.streetNum(config.numstreet);
+            this.companySize(config.sizecompany);
+            $('select').on('change', function () {
```
Comment (bricht) [inline Devall/Tabatadze/view/frontend/web/js/details.js:14]: what if another select is added to that template at some point? You need to be more specific.

File: Devall/Tabatadze/view/frontend/web/js/details.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+            function setCompanyDetails(details) {
+                feature.country(details.country);
+                feature.street(details.street);
+                feature.streetNum(details.street_number);
+                feature.companySize(details.size);
+
+            }
+        },
+        country: ko.observable(""),
```
Comment (bricht) [inline Devall/Tabatadze/view/frontend/web/js/details.js:35]: declare there at the top of your component for clarity, you can look at other js components in vendor to see some examples.

File: Devall/Tabatadze/view/frontend/web/js/details.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        street: ko.observable(""),
+        streetNum: ko.observable(""),
+        companySize: ko.observable(""),
+
+        getCountry: function () {
```
Comment (bricht) [inline Devall/Tabatadze/view/frontend/web/js/details.js:40]: docBlock please

File: Devall/Tabatadze/view/frontend/web/js/details.js [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    return Component.extend({
+        initialize: function (config) {
+            var feature = this;
+            this._super();
+            this.country(config.countrycompany);
+            this.street(config.streetcompany);
+            this.streetNum(config.numstreet);
+            this.companySize(config.sizecompany);
+            $('select').on('change', function () {
+                var valueSelected = this.value - 1;
```
Comment (bricht) [inline Devall/Tabatadze/view/frontend/web/js/details.js:15]: Why the -1? What is it?

**Review by bricht — CHANGES_REQUESTED** [review]

File: Devall/Tabatadze/view/frontend/web/js/details.js (hunk 23 above)
Comment (davidtabat) [reply Devall/Tabatadze/view/frontend/web/js/details.js:15]: i'm doing get request on all companies, Array indexes starts with 0 company ids with 1
image
